Fixed #24886 — Added Transform.process_lhs().

Samriddha9619 opened the pull request:
ticket-24886

PR Description
This PR introduces a process_lhs() method to the Transform base class, aligning it with the Lookup API, which already provides both process_lhs() and process_rhs(). This improves consistency in SQL generation and makes transform behavior easier to reason about.

Note: This patch does not refactor any existing transforms. The method is introduced as a structural enhancement to support future consistency and maintainability. Refactoring built-in transforms (e.g. Upper, Lower) can follow once this pattern is approved and validated.

There are no functional changes and no new tests added, as this is a scoped internal refactor.

 Branch Description
This branch introduces process_lhs() to the Transform base class to match the structure of Lookup, improving consistency and maintainability in SQL generation logic.

Checklist
[ ] This PR targets the main branch.

[ ] The commit message is written in past tense, mentions the ticket number, and ends with a period.

[ ] I have checked the "Has patch" ticket flag in the Trac system.

[ ] I have added or updated relevant tests. (Not applicable for this structural refactor)

[ ] I have added or updated relevant docs, including release notes if applicable.

[ ] I have attached screenshots in both light and dark modes for any UI changes

sarahboyce commented:
Thank you for the PR :+1: 
I have a couple of thoughts:
- we might want to think about documenting this in the `Transform` docs (https://docs.djangoproject.com/en/5.2/ref/models/lookups/#django.db.models.Transform) or the `Transform` examples (https://docs.djangoproject.com/en/5.2/howto/custom-lookups/#a-transformer-example)
- I believe the point of this method would then to update `lhs, lhs_params = compiler.compile(self.lhs)` within `as_sql()` methods which inherit `Transform` to be `lhs, lhs_params = self.process_lhs(compiler, connection)`

Samriddha9619 commented:
@sarahboyce I’ve added process_lhs() to the Lookup class and updated the transformer example in custom-lookups.txt to use it in as_sql(). I also added .. versionadded:: 6.1 to the API docs in ref/models/lookups.txt to mark the method addition. I noticed a few different documentation placement options were mentioned, and I tried to choose what felt most appropriate—but I wasn’t entirely sure if this was the right level of detail or location. Please let me know if you'd prefer a deeper example, a different doc placement, or anything else. Happy to adjust!

sarahboyce commented on docs/ref/models/lookups.txt:
```diff
@@ -237,6 +237,8 @@ the following methods:
         can be used for compiling vendor specific SQL. If ``lhs`` is not
         ``None``, use it as the processed ``lhs`` instead of ``self.lhs``.
 
+        .. versionadded:: 6.1
+
```
```suggestion
```
This existed for Lookup prior to this change so we don't add this here

What I meant would have been something like this:
```diff
--- a/docs/ref/models/lookups.txt
+++ b/docs/ref/models/lookups.txt
@@ -180,6 +180,13 @@ the following methods:
         :class:`~django.db.models.Field` instance. By default is the same as
         its ``lhs.output_field``.
 
+    .. method:: process_lhs(compiler, connection, lhs=None)
+
+        .. versionadded:: 6.1
+
+        Same as :meth:`~Lookup.process_lhs`. Returns a tuple
+        ``(lhs_string, lhs_params)``, as returned by ``compiler.compile(lhs)``.
```

sarahboyce commented on docs/howto/custom-lookups.txt:
```diff
@@ -122,6 +122,14 @@ SQL function ``ABS()`` to transform the value before comparison::
       lookup_name = "abs"
       function = "ABS"
 
+      def as_sql(self, compiler, connection):
+          lhs, lhs_params = self.process_lhs(compiler, connection)
+          return "%s(%s)" % (self.function, lhs), lhs_params
+
+This override uses ``process_lhs()`` to safely compile the left-hand side of
+the expression. This is recommended when customizing SQL generation in
+``Transform.as_sql()`` methods.
+
```
```suggestion
```
I think let's not complicate this example unnecessarily
Let's only add an example if we have a clean way to add it

Samriddha9619 replied:
I want to sincerely apologize for the confusion and misunderstandings in this PR; I realize now that I didn’t fully grasp the points you were trying to convey and misunderstood the scope and implications of the changes I proposed, and I truly appreciate the time and patience you’ve shown in guiding me despite my mistakes—I’ll take a step back to better study the internals and discussions before moving forward, and I’m sorry for any extra work or noise this caused.

charettes commented:
@sarahboyce I think the idea still has merit but as you alluded to a full review of `Transform` and `Lookup` used as transform has to made to see how/if they can take advantage of this. In other words, this ticket is less about adding the method but more about seeing how much it can simplify the maintenance of these objects and thus be valuable as an _extra_ hook for users who write their own lookup and transforms.

Samriddha9619 commented:
@sarahboyce I want to sincerely apologize for the confusion and misunderstandings in this PR; I realize now that I didn’t fully grasp the points you were trying to convey and misunderstood the scope and implications of the changes I proposed, and I truly appreciate the time and patience you’ve shown in guiding me despite my mistakes—I’ll take a step back to better study the internals and discussions before moving forward, and I’m sorry for any extra work or noise this caused.

charettes commented:
@Samriddha9619 no need to apologize, I don't think the ticket captured this intent properly and the work you've so far is great.

There's always a the risk when picking up old tickets that they require more work than they might look like and this one is no exception.

Samriddha9619 commented:
@sarahboyce Thanks for pointing out the other Transform subclasses and test examples—those definitely seem worth exploring. I scoped this patch narrowly to avoid overreach, but I’d be happy to look into extending the refactor to RasterBandTransform, Trunc*, and others if you think it’s worthwhile.
If this feels too marginal or not worth expanding, I’m okay stepping back from the patch. But if there’s future direction or guidance, I’d love to keep contributing.

charettes commented:
@Samriddha9619 thanks for your work so far, it it definitely in the right direction!

In my opinion there is no harm in trying to adapt as many `Transform` subclasses to the new pattern given we are going to introduce it. It might teach us a thing or two about why this approach needs adjustments and potentially make use of the currently unused `lhs=None` kwarg.

As long as they are added in separate commit it should make it easy for the mergers to cherry-pick the changes that makes the most sense (I suspect they all will).

Let me know if that answer your questions.

Samriddha9619 commented:
Hi @charettes, thank you for the guidance!
I've now updated the PR to include the additional Transform subclasses you and @sarahboyce mentioned. The changes are organized in separate commits as you suggested, covering:

Extract, TruncBase, TruncDate, TruncTime (datetime functions)
RasterBandTransform (GIS)
Test examples in custom_lookups/tests.py

I was careful to only modify classes that inherit from Transform and avoided touching CombinedExpression or other non-Transform classes.
I apologize for the delay in responding—I appreciate your patience. Please let me know if there are any other adjustments needed or if this approach looks good for review. Happy to iterate further!

charettes commented on django/db/models/lookups.py:
```diff
@@ -215,6 +215,10 @@ class Transform(RegisterLookupMixin, Func):
+    def process_lhs(self, compiler, connection, lhs=None):
+        lhs = lhs or self.lhs
+        return compiler.compile(lhs)
+
```
For those who were wondering `lhs=None` is completely unused internally for `Transform` subclasses.

It's done this way to maintain symmetry with `Lookup.process_lhs` which was the main motivating factor for this ticket in the first place.

charettes reviewed (approved):
Thanks for the adjustments @Samriddha9619 and no worries about the delay, there is no rush here.

I searched for all `Transform` usage and I couldn't find any other instance that needs adjustments so that LGTM!

raffaellasuardini commented:
Hello @Samriddha9619,
Me and @ontowhee looked at the PR and would like to suggest an update to the documentation as [Sarah suggested](https://github.com/django/django/pull/19885#discussion_r2388122658) with having a similar result that already is in the [Lookups](https://docs.djangoproject.com/en/6.0/ref/models/lookups/#lookup-reference) but this time placed in the `Transform`. Here is addition to the `Transform`.  

```diff
--- a/docs/ref/models/lookups.txt
+++ b/docs/ref/models/lookups.txt
@@ -180,6 +180,13 @@ the following methods:
         :class:`~django.db.models.Field` instance. By default is the same as
         its ``lhs.output_field``.
 
+    .. method:: process_lhs(compiler, connection, lhs=None)
+
+        .. versionadded:: 6.1
+
+        Same as :meth:`~Lookup.process_lhs`. Returns a tuple
+        ``(lhs_string, lhs_params)``, as returned by ``compiler.compile(lhs)``.
```

ontowhee commented:
> I noticed that `KeyTransform` has a `preprocess_lhs()` which might be interesting to look at also

@raffaellasuardini and I took a look into [KeyTransform in json.py](https://github.com/django/django/blob/main/django/db/models/fields/json.py#L488) which has the `preprocess_lhs()` function that Sarah mentioned. We do not recommend any changes here. The reason is, the value being compiled is not the immediate lhs value, but rather a previous value in the chain of transforms.

It could potentially make this call by passing in `previous` as the `lhs` argument, as below,

```diff
--- a/django/db/models/fields/json.py
+++ b/django/db/models/fields/json.py
@@ -359,7 +359,7 @@ class KeyTransform(Transform):
         while isinstance(previous, KeyTransform):
             key_transforms.insert(0, previous.key_name)
             previous = previous.lhs
-        lhs, params = compiler.compile(previous)
+        lhs, params = self.process_lhs(compiler, connection, lhs=previous)
         if connection.vendor == "oracle":
             # Escape string-formatting.
             key_transforms = [key.replace("%", "%%") for key in key_transforms]
```

However, this is more confusing than just calling `compiler.compile(previous)` directly. As Simon mentioned,

> lhs=None` is completely unused internally for `Transform` subclasses.

and attempting to use it here would break that convention.
